### PR TITLE
Optimization: do not use network transport for local spawnAsync.

### PR DESCRIPTION
**Pull request by qnikst**

### CODE REVIEW

**Comment by qnikst** [comment]
This is PR for a tweag patches, if patch is fine, I'll prepare PR to upstream

File: src/Control/Distributed/Process/Internal/Primitives.hs [hunk 1]
```diff
@@ -982,7 +982,9 @@ expectTimeout n = receiveTimeout n [match return]
 spawnAsync :: NodeId -> Closure (Process ()) -> Process SpawnRef
 spawnAsync nid proc = do
   spawnRef <- getSpawnRef
-  sendCtrlMsg (Just nid) $ Spawn proc spawnRef
+  node     <- getSelfNode
+  let dest  = if nid == node then Nothing else Just nid
+  dest `seq` sendCtrlMsg dest $ Spawn proc spawnRef
```
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Primitives.hs]: Shouldn't this be:
```Haskell
if nid == here then
  sendCtrlMsg Nothing $ force $ Spawn proc spawnRef
else
  sendCtrlMsg (Just nid) $ Spawn proc spawnRef
```
Forcing is only needed in the local case, and it is needed to ensure any errors in computing the expression are thrown in the sender thread.

File: src/Control/Distributed/Process/Internal/Primitives.hs [hunk 2]
```diff
@@ -982,7 +983,10 @@ expectTimeout n = receiveTimeout n [match return]
 spawnAsync :: NodeId -> Closure (Process ()) -> Process SpawnRef
 spawnAsync nid proc = do
   spawnRef <- getSpawnRef
-  sendCtrlMsg (Just nid) $ Spawn proc spawnRef
+  node     <- getSelfNode
+  if nid == node
+    then sendCtrlMsg Nothing $ Spawn (DS.force proc) spawnRef
```
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Primitives.hs]:
```Haskell
sendCtrlMsg Nothing $! Spawn (DS.force proc) spawnRef
```
?

File: src/Control/Distributed/Process/Internal/Types.hs [hunk 3]
```diff
@@ -559,7 +559,9 @@ data NCMsg = NCMsg
   { ctrlMsgSender :: !Identifier
   , ctrlMsgSignal :: !ProcessSignal
   }
-  deriving Show
+  deriving (Show, Generic)
+
+instance NFData NCMsg
```
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Types.hs]: Unnecessary now?

Comment (qnikst) [inline src/Control/Distributed/Process/Internal/Types.hs]: I'd say there is no reason to not have this instance.

File: src/Control/Distributed/Process/Internal/Primitives.hs (hunk 2 above)
Comment (qnikst) [inline src/Control/Distributed/Process/Internal/Primitives.hs]: yes

File: src/Control/Distributed/Process/Internal/Types.hs (hunk 3 above)
Comment (facundominguez) [inline src/Control/Distributed/Process/Internal/Types.hs]: Then make it into another commit/PR and maybe motivate it?

**Comment by facundominguez** [comment]
LGTM

**Comment by facundominguez** [comment]
Perhaps we could have a test calling `spawnAsync here (error "bang")`.

**Comment by qnikst** [comment]
Closing?

**Comment by facundominguez** [comment]
Yep, it is not needed in tweag-patchset.
